awx/main/scheduler/task_manager_instances.py

```diff
@@ -0,0 +1,252 @@
+
+    def __repr__(self):
+        return self.__str__()
+
+
+class PrioritizedNodes:
+    def __init__(self):
+        self.nodes = set()
+
+    def add(self, node):
+        self.nodes.add(node)
+
+    def __getitem__(self, hostname):
+        if hostname not in self.nodes:
+            raise KeyError(f"No control node with hostname {hostname} found")
+        for node in self.nodes:
+            if node == hostname:
+                return node
+
+    def update(self, node):
+        """The __hash__ function for PrioritizableNodes just compares hostnames.
+
+        We can add/remove the node to update object in PrioritizedNodes to have right capacity.
+        """
+        self.nodes.remove(node)
+        self.nodes.add(node)
+
+    def best_node(self):
+        heap = list(self.nodes)
+        heapq.heapify(heap)
+        return heapq.nlargest(1, heap).pop()
+
+    def __str__(self):
+        return f"{[n for n in self.nodes]}"
+
+
+class TaskManagerInstances:
+    def __init__(self):
+        realinstances = Instance.objects.filter(hostname__isnull=False, enabled=True).exclude(node_type='hop').prefetch_related('rampart_groups')
+        self.instances_partial = dict()
+        self.control_nodes = PrioritizedNodes()
+
+        for instance in realinstances:
+            self.instances_partial[instance.hostname] = SimpleNamespace(
+                obj=instance,
+                node_type=instance.node_type,
+                remaining_capacity=instance.remaining_capacity,
+                capacity=instance.capacity,
+                jobs_running=instance.jobs_running,
+                hostname=instance.hostname,
+                instance_groups=[ig.name for ig in instance.rampart_groups.all()],
+            )
+            if instance.node_type in ('control', 'hybrid'):
+                self.control_nodes.add(PrioritizableNode(remaining_capacity=instance.remaining_capacity, hostname=instance.hostname))
+
+    def __getitem__(self, key):
+        return self.instances_partial[key]
+
+    @staticmethod
+    def find_largest_idle_instance(instances, capacity_type='execution'):
+        largest_instance = None
+        for i in instances:
+            if i.node_type not in (capacity_type, 'hybrid'):
+                continue
+            if i.jobs_running == 0:
+                if largest_instance is None:
+                    largest_instance = i
+                elif i.capacity > largest_instance.capacity:
+                    largest_instance = i
+        return largest_instance
+
+    @staticmethod
+    def fit_task_to_most_remaining_capacity_instance(task, instances):
+        instance_most_capacity = None
+        for i in instances:
+            if i.node_type not in (task.capacity_type, 'hybrid'):
+                continue
+            if i.remaining_capacity >= task.task_impact and (
+                instance_most_capacity is None or i.remaining_capacity > instance_most_capacity.remaining_capacity
```

Comment:
Instead of considering `i.remaining_capacity >= task.task_impact` as the criteria here, I think you should consider the criteria to be `i.remaining_capacity >= task.task_impact + settings.AWX_CONTROL_NODE_TASK_IMPACT` for hybrid nodes.

This should be actually pretty easy to do, although you'll be adding a little bit more python code to this method. This avoids any situation where we "fit" something but it won't actually fit because the control capacity consumption pushes it over.

This is a key change that I alluded to but probably didn't communicate effectively. If it doesn't make sense, let me know. It's really important for what needs to happen here.